chore(github): migrate to github actions from travis

Langleu opened the pull request:
related to [INFRA-2867](https://jira.camunda.com/browse/INFRA-2867)

The previous Travis setup of:

```
env:
  - DISTRO=tomcat EE=false
  - DISTRO=wildfly EE=false
  - DISTRO=run EE=false

jobs:
  include:  # do not run EE builds for forks (they don't have access to EE credentials)
    - if: head_repo = repo
      env: DISTRO=tomcat EE=true
    - if: head_repo = repo
      env: DISTRO=wildfly EE=true
    - if: head_repo = repo
      env: DISTRO=run EE=true

script:
  - ./pipeline.sh # contained release.sh that checked everytime whether certain conditions apply
```

Split up the current Travis setup into 2 workflows:
1. runs on every commit the CE Matrix and merge if (next/version) as head ref.
2. runs on every PR on top of the EE Matrix in case the PR comes from within the repo

Is dependent on the following PR to add the secrets: https://github.com/camunda/infra-core/pull/3430
Removed the docker hub mail (`DOCKER_HUB_EMAIL`) as it is not referenced anywhere in this repo and doesn't seem to be a docker default env.

EE tests are not failing even with secrets not supplied. The build itself fails but continues the tests with the CE version (looks for `...:tomcat`, which is supposed to be built but is an actual docker hub image. (Same would happen on Travis)

Langleu committed:
chore(github): migrate to github actions from travis

related to INFRA-2867

lwille commented on .github/workflows/build-and-test.yml:
```diff
@@ -0,0 +1,21 @@
+  build-and-test:
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        include:
+          - DISTRO: "tomcat"
+            EE: "false"
+          - DISTRO: "wildfly"
+            EE: "false"
+          - DISTRO: "run"
+            EE: "false"
```
There's not really a need for a complex matrix if `EE` is always false.
```suggestion
      matrix:
        DISTRO: [tomcat, wildfly, run]
```

Langleu replied:
thanks, good idea!

lwille commented on .github/workflows/build-and-test.yml:
```diff
@@ -0,0 +1,21 @@
+        run: ./pipeline.sh
+        env:
+          DISTRO: ${{ matrix.DISTRO }}
+          EE: ${{ matrix.EE }}
```
I'd suggest adding a conditional `run` step here (reason in other comment), which will only get executed for `next` and tags matching the semver pattern:


```suggestion
          EE: "false"
      - name: Publish image
        if: {{ github.ref == "refs/heads/next" || startsWith(github.ref, "refs/tags/7.") }}
        run: ./release.sh
        env:
          DISTRO: ${{ matrix.DISTRO}}
          EE: "false"
          DOCKER_HUB_PASSWORD: ${{ secrets.DOCKER_HUB_PASSWORD }}
          DOCKER_HUB_USERNAME:  ${{ secrets.DOCKER_HUB_USERNAME }}
```

Langleu replied:
you're right, makes sense! Didn't see prior that it required the image to already been built.

lwille commented on .github/workflows/build-and-test-ee.yml:
```diff
@@ -0,0 +1,24 @@
+        run: ./pipeline.sh
+        env:
+          DISTRO: ${{ matrix.DISTRO }}
+          EE: ${{ matrix.EE }}
```
```suggestion
          EE: true
```

lwille commented on .github/workflows/build-and-test-ee.yml:
```diff
@@ -0,0 +1,24 @@
+    if: github.event.pull_request.head.repo.full_name == github.repository
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        include:
+          - DISTRO: "tomcat"
+            EE: "true"
+          - DISTRO: "wildfly"
+            EE: "true"
+          - DISTRO: "run"
+            EE: "true"
```
There's not really a need for a complex matrix if EE is always `true` for this workflow.

```suggestion
        DISTRO: [tomcat, wildfly, run]
```

lwille commented on .github/workflows/publish-ce-image.yml:
```diff
@@ -0,0 +1,27 @@
+name: Publish CE image
+on:
+  push:
+    branches:
+      - next
+      - 7.[0-9]+ # not real regex (https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet)
+jobs:
+  publish-ce-image:
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        include:
+          - DISTRO: "tomcat"
+            EE: "false"
+          - DISTRO: "wildfly"
+            EE: "false"
+          - DISTRO: "run"
+            EE: "false"
+    steps:
+      - uses: actions/checkout@v2
+      - name: Release CE docker image
+        run: ./release.sh
+        env:
+          DISTRO: ${{ matrix.DISTRO }}
+          EE: ${{ matrix.EE }}
+          DOCKER_HUB_PASSWORD: ${{ secrets.DOCKER_HUB_PASSWORD }}
+          DOCKER_HUB_USERNAME:  ${{ secrets.DOCKER_HUB_USERNAME }}
```
I think this should be part of the `build-and-test` (for CE) workflow:

Because this is a separate workflow, it doesn't have access to the images that were built in the `build-and-test` workflow - they were only created on that runner and are deleted once the run completes.


```suggestion
```

Langleu committed:
chore(github): rework based on PR suggestions

related to INFRA-2867

Langleu committed:
chore(README): remove Travis badge

maxdanilov commented on .github/workflows/README.md:
```diff
@@ -0,0 +1,28 @@
+# GitHub Actions
+
```
What do you think about making this document not transient?
Currently, it heavily builds on the fact this repo has been using Travis CI.
In a few weeks time, nobody would remember that, and all potential fixes to the GHA workflows would be done without having Travis in mind. 
I would move the `The previous Travis setup of` section of the document to the PR description, and keep the list of GHA actions in this doc, explaining what they do, without referring to Travis at all.

maxdanilov reviewed (commented):
Pretty cool, awesome to see the GHA in action 🎉
Just leaving a small comment for the readme.

maxdanilov commented:
Would these workflows need to be backported to all `7.x` branches?

Langleu commented:
@maxdanilov, yes pretty much till 7.8 as that is the last one that still receives patch support but will just in case check-in with Thorben as in theory it overwrites the image every time (which was the case with Travis as well) and CE releases are not patched ... only EE, which makes it kinda pointless.

Langleu committed:
chore(github): rewrite readme about workflows

related to INFRA-2867

Langleu committed:
chore(release.sh): only publish CE image if not available on Docker

Langleu commented:
@maxdanilov, added your changes and reworked the `release.sh` a bit.
Backport till 7.8 is required and as CE images are only released once, I've added the following line to the `release.sh`

```
# check whether the CE image for distro was already released and exit in that case
if [ $(docker manifest inspect $IMAGE:${DISTRO}-${VERSION} > /dev/null ; echo $?) == '0' ]; then
    echo "Not pushing already released CE image"
    exit 0
fi
```

This will check whether the version exists already for distro and version and not re-release in that case.
E.g. 7.16 wouldn't re-release and 7.17 would be able to release as well as snapshots till 7.17 is finally released.

Backport will happen after merged to next.
